### PR TITLE
Docker in Local Dev

### PR DESCRIPTION
**Pull request by ollyjshaw**
This PR:
* allows site developers to run more fine grained tasks (eg not have to restart docker services to restart awestruct)
* Makes the features test more reliable
* Adds a convenient 'run the stack' option
* Adds guard and unit testing
* Re-write of the readme to show how to get a dev setup
* Conversion of options to tasks (a refactor that will help future CI jobs)

### CODE REVIEW

**Comment by paulrobinson** [comment]
@LightGuard can you review? I'll take a look on Monday too.

File: README.md [hunk 1]
```diff
@@ -1,34 +1,98 @@
...
+
+Skip to the [Site Build Setup](#site_build_setup) section if you don't want to use Docker.
+###Utilities
+You should be running a bash shell (Linux or OSX) and you will require: git and curl.
+###Docker and Docker compose
+Follow the instructions to install the latest docker for your system [here] (https://docs.docker.com/installation/). It's unlikely you will want to use the packages provided by your system as they will be too far out of date.
+
+Follow the instructions to install the latest docker-compose [here] (https://github.com/docker/compose/releases)
+### Basic Ruby install
+In this project docker and docker-compose are managed through the ruby script found at `_docker/control.rb`. In order to run this you will require ruby 2.1.2. The following instructions install ruby via rbenv. You can use other methods, but your mileage may vary.
```
Comment (LightGuard) [inline README.md]: Ruby 2.1.2 should be the base line, but it runs fine on any 2.1 or higher version of ruby.

File: README.md [hunk 2]
```diff
@@ -1,34 +1,98 @@
...
+Download the Searchisko data dump from [here](https://github.com/redhat-developer/dcp-dumps/raw/master/searchisko.sql.zip) and copy to `_docker/searchisko/overlay/searchisko.sql.zip`
+
+Add the host `docker` to your `/etc/hosts` file. If you are building on Linux, set the IP address to `127.0.0.1`. If you are on a Mac and thus using Boot2Docker, you will need to set the IP address to that of your Boot2Docker image. You can discover this IP address by running `boot2docker ip`
+
+Run `bundle install` from within the `_docker` directory to download the necessary ruby gems.
+
+###Boot2Docker setup (OSX only)
+If you are running on Mac you will need to follow the steps below in _Edit your boot2docker DNS setup_ to setup configuration for the Red Hat DNS and VPN.
+
+If you are working on OSX or widows via boot to docker, then set the environment variable DOCKER_SSL_VERIFY to false. E.g `export DOCKER_SSL_VERIFY=false`. See here as to why https://github.com/swipely/docker-api/issues/202
```
Comment (LightGuard) [inline README.md]: windows -> Windows

File: README.md [hunk 3]
```diff
@@ -1,34 +1,98 @@
...
-6. If you are running on Mac you will need to follow the steps below in _Edit your boot2docker DNS setup_ to setup configuration for the Red Hat DNS and VPN.
-7. If you need to override the default DOCKER_HOST this can be done via the `-d` flag. This should not be necessary in most cases.
-8. If you are working on OSX or widows via boot to docker, then set the environment variable DOCKER_SSL_VERIFY to false. E.g `export DOCKER_SSL_VERIFY=false`. See here as to why https://github.com/swipely/docker-api/issues/202
-9. Run the following commands to build the images and start the containers:
+` bundle exec control.rb --run-the-stack`
 
-        ./control.rb -br
+This will take awhile the first time. The starts all required services and then  runs awestruct in preview mode. The script won't finish until you stop it with `CTRL+C`. At the start of the build the script will output the ports the services are listening on for access outside of docker. Typically you'll only need to worry about awestruct and searchisko ports. Those will be available on host `docker` and the corresponding port for that service.
```
Comment (LightGuard) [inline README.md]: The starts -> This starts ?

File: README.md [hunk 4]
```diff
@@ -1,34 +1,98 @@
...
 
-This will take awhile the first time. On subsequent builds you do not need to pass the `-b` flag unless you need to rebuild the docker containers. The script will output some information when everything is ready to go and you'll see the familiar awestruct build as you near the end. The script runs awestruct in preview mode, so the script won't finish until you stop it with `CTRL+C`. At the start of the build the script will output the ports the services are listening on for access outside of docker. Typically you'll only need to worry about awestruct and searchisko ports. Those will be available on host `docker` and the corresponding port for that service.
+NOTE: When `preview` is run, you may see errors from guard/listen about a folder being watched already, as far I as I can tell this is harmless and you can ignore those.
```
Comment (LightGuard) [inline README.md]: Correct, those are harmless. It comes from guard watching at a higher level than the file and also trying to watch the file.

File: README.md [hunk 5]
```diff
@@ -1,34 +1,98 @@
...
 
-NOTE: When `preview` is run, you may see erorrs from guard/listen about a folder being watched already, as far I as I can tell this is harmless and you can ignore those.
+###Important Control.rb commands
+To build and start all services:
+
+`bundle exec ./control.rb --run-the-stack`
```
Comment (LightGuard) [inline README.md]: Is there a single arg "run-the-stack" option?

File: _docker/docker-compose.yml.erb [hunk 6]
```diff
@@ -29,9 +29,9 @@ awestruct:
     - github_token
     - drupal_user
     - drupal_password
-    - AWESTRUCT_HOST_PORT
-    - SEARCHISKO_HOST_PORT
-    - DRUPAL_HOST_PORT
+    - AWESTRUCT_HOST_PORT=<%=ENV['AWESTRUCT_HOST_PORT']%>
+    - SEARCHISKO_HOST_PORT=<%=ENV['SEARCHISKO_HOST_PORT']%>
+    - DRUPAL_HOST_PORT=<%=ENV['DRUPAL_HOST_PORT']%>
```
Comment (LightGuard) [inline _docker/docker-compose.yml.erb]: These aren't required as docker-compose will use the ENV variables of the same name from the parent process, but this doesn't hurt.

Comment (ollyjshaw) [inline _docker/docker-compose.yml.erb]: Thanks for the comments so far. 

For this particular one, the reason that these environments are set is to allow the ports to remain consistent if someone is restarting awestruct. All the other env variables are ok to set again in the environment. The ports would be different as the services will already be running. 

I've probably not explained it too well,  but we need those ports set :-)

Comment (LightGuard) [inline _docker/docker-compose.yml.erb]: Got it. No problem

On Friday, October 2, 2015, Oliver Shaw <notifications@github.com> wrote:

> In _docker/docker-compose.yml.erb
> <https://github.com/redhat-developer/developers.redhat.com/pull/448#discussion_r41039651>
> :
>
> > @@ -29,9 +29,9 @@ awestruct:
> >      - github_token
> >      - drupal_user
> >      - drupal_password
> > -    - AWESTRUCT_HOST_PORT
> > -    - SEARCHISKO_HOST_PORT
> > -    - DRUPAL_HOST_PORT
> > +    - AWESTRUCT_HOST_PORT=<%=ENV['AWESTRUCT_HOST_PORT']%>
> > +    - SEARCHISKO_HOST_PORT=<%=ENV['SEARCHISKO_HOST_PORT']%>
> > +    - DRUPAL_HOST_PORT=<%=ENV['DRUPAL_HOST_PORT']%>
>
> Thanks for the comments so far.
>
> For this particular one, the reason that these environments are set is to
> allow the ports to remain consistent if someone is restarting awestruct.
> All the other env variables are ok to set again in the environment. The
> ports would be different as the services will already be running.
>
> I've probably not explained it too well, but we need those ports set :-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/redhat-developer/developers.redhat.com/pull/448/files#r41039651>
> .
>


-- 
Sent from Gmail Mobile

File: README.md (hunk 5 above)
Comment (ollyjshaw) [inline README.md]: yes, I've added it to hopefully make it a bit easier for someone to just get it going locally.

File: README.md (hunk 2 above)
Comment (ollyjshaw) [inline README.md]: I've removed this altogether as the instructions don't support a windows install currently.

File: README.md [hunk 7]
```diff
@@ -1,34 +1,98 @@
...
-4. Add the host `docker` to your `/etc/hosts` file. If you are building on Linux, set the IP address to `127.0.0.1`. If you are on a Mac and thus using Boot2Docker, you will need to set the IP address to that of your Boot2Docker image. You can discover this IP address by running `boot2docker ip`
-5. Run `bundle install` from within the `_docker` directory to download the necessary ruby gems.
-6. If you are running on Mac you will need to follow the steps below in _Edit your boot2docker DNS setup_ to setup configuration for the Red Hat DNS and VPN.
-7. If you need to override the default DOCKER_HOST this can be done via the `-d` flag. This should not be necessary in most cases.
-8. If you are working on OSX or widows via boot to docker, then set the environment variable DOCKER_SSL_VERIFY to false. E.g `export DOCKER_SSL_VERIFY=false`. See here as to why https://github.com/swipely/docker-api/issues/202
-9. Run the following commands to build the images and start the containers:
+` bundle exec control.rb --run-the-stack`
```
Comment (paulrobinson) [inline README.md]: ./control.rb

File: README.md [hunk 8]
```diff
@@ -1,34 +1,98 @@
...
 
-        ./control.rb -br
+This will take awhile the first time. This starts all required services and then  runs awestruct in preview mode. The script won't finish until you stop it with `CTRL+C`. At the start of the build the script will output the ports the services are listening on for access outside of docker. Typically you'll only need to worry about awestruct and searchisko ports. Those will be available on host `docker` and the corresponding port for that service.
```
Comment (paulrobinson) [inline README.md]: awhile -> a while

**Pull request merged**